File: packages/sdk-go/examples/file-upload.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"os"
	"path/filepath"

	stagehand "github.com/browserbase/stagehand/packages/sdk-go"
)

func main() {
	if err := run(context.Background()); err != nil {
		log.Fatal(err)
	}
}

func run(ctx context.Context) (err error) {
	directory, err := os.MkdirTemp("", "stagehand-upload-")
	if err != nil {
		return err
	}
	defer os.RemoveAll(directory)
	filePath := filepath.Join(directory, "hello.txt")
	if err := os.WriteFile(filePath, []byte("hello from Go"), 0o600); err != nil {
		return err
	}

	browser, err := stagehand.LaunchLocalBrowser(
		ctx,
		&stagehand.LocalBrowserLaunchOptions{Headless: true},
	)
	if err != nil {
		return err
	}
	defer func() { err = errors.Join(err, browser.Close(ctx)) }()

	client, err := stagehand.Create(ctx, stagehand.CreateOptions{Browser: browser})
	if err != nil {
		return err
	}
	defer func() { err = errors.Join(err, client.Close(ctx)) }()

	browserContext, err := client.Context()
	if err != nil {
		return err
	}
	page, err := browserContext.ActivePage(ctx)
	if err != nil {
		return err
	}
	if page == nil {
		return errors.New("Stagehand initialized without an active page")
	}
	if err := page.Goto(ctx, `data:text/html,<input id="upload" type="file">`, nil); err != nil {
		return err
	}
	if err := page.Locator("#upload").SetInputFiles(ctx, stagehand.FilePath(filePath)); err != nil {
		return err
	}

	raw, err := page.Evaluate(ctx, `(async () => {
      const file = document.querySelector('#upload').files[0];
      return file ? { name: file.name, text: await file.text() } : null;
    })()`)
	if err != nil {
		return err
	}
	var uploaded struct {
		Name string `json:"name"`
		Text string `json:"text"`
	}
	if err := json.Unmarshal(raw, &uploaded); err != nil {
		return err
	}
	if uploaded.Name != "hello.txt" || uploaded.Text != "hello from Go" {
		return fmt.Errorf("unexpected uploaded file: %+v", uploaded)
	}
	fmt.Printf("%+v\n", uploaded)
	return nil
}

File: packages/sdk-ruby/examples/file_upload.rb
# frozen_string_literal: true

# Ruby port of packages/sdk-python/examples/file_upload.py. No LLM involved —
# this exercises locator.set_input_files + page.evaluate only.
#
#   ruby examples/file_upload.rb                # local Chrome
#   ruby examples/file_upload.rb --browserbase  # Browserbase session

require "tmpdir"

require_relative "../lib/stagehand"

Dir.mktmpdir("stagehand-upload-") do |directory|
  file_path = File.join(directory, "hello.txt")
  File.write(file_path, "hello from Ruby")

  browser =
    if ARGV.include?("--browserbase")
      puts "Creating a Browserbase session..."
      Stagehand::Browserbase.launch(api_key: ENV.fetch("BROWSERBASE_API_KEY"))
    else
      puts "Launching local Chrome..."
      Stagehand::LocalBrowser.launch(headless: ENV["HEADED"].nil?)
    end

  begin
    stagehand = Stagehand.create(browser: browser, log_level: "warn")
    begin
      page = browser.context.active_page || browser.context.pages.first

      page.goto('data:text/html,<input id="upload" type="file">')
      page.locator("#upload").set_input_files(file_path)

      uploaded = page.evaluate(<<~JS)
        (async () => {
          const file = document.querySelector('#upload').files[0];
          return file ? { name: file.name, text: await file.text() } : null;
        })()
      JS
      expected = { "name" => "hello.txt", "text" => "hello from Ruby" }
      raise "Unexpected uploaded file: #{uploaded.inspect}" unless uploaded == expected
      puts uploaded.inspect
    ensure
      stagehand.close
    end
  ensure
    browser.close
  end
end
